### DISCUSSION TITLE
Where and how to communicate Dagger's API changes?

### DISCUSSION THREAD

**Discussion by vito** [dagger/dagger#3472] [Help]
We're on the cusp of releasing the first semi-stable version of Dagger's new API + Go SDK. Up until now we've been changing everything constantly, making breaking changes every day, but once we have shipped an API that other folks code against we'll need to be more thoughtful and communicative.

This discussion is just to figure out how and where we should be communicating those changes.

Some points in no particular order:

* Changes to the core API imply changes to every SDK.
* Changes to any SDK imply changes to end-user code (including extensions in the future).
* Changes to SDKs for compiled languages will be _much_ easier to notice and fix than changes in interpreted languages.
* Additive changes can probably just be new features in release notes, but we might want to communicate them loudly too if they fix particularly common pain points.
* Breaking changes need to be communicated loudly and clearly.
* We're pre-1.0, so we need something more "loud and clear" than a major version bump since it's perfectly normal for v0.X.0 to have breaking changes.
* Users should be able to see not only see breaking changes, but the justification for them and how to migrate their code.

Some prior art:

* Generating release notes based on merged PRs is really easy and has a very low maintenance burden as it completely eliminates a manual step from the release process. GitHub has this feature built-in with configurable sections based on labels ([example](https://github.com/vito/bass/blob/main/.github/release.yml)). But if we do this we might want to be more thoughtful in our PR descriptions; they're also effectively the release notes so they'd need to say the "why" and "how to migrate."
* Some projects post a single pinned + locked issue for announcing breaking changes ([example](https://github.com/hrsh7th/nvim-cmp/issues/231)). This is pretty clever because people can subscribe to it and maintainers have a rich posting/editing experience. Granted, I think this is more necessary for projects that don't actually have a release process (e.g. vim plugins which are typically YOLO'd), but the pattern is pretty cute nonetheless.
* Manually curated `CHANGELOG.md` files is a pretty common practice but they're difficult to subscribe to and typically involve a manual separate authoring step. Separate files _might_ be easier to follow than GitHub releases though; GitHub releases can get kind of confusing once you're shipping multiple things from one repo.

If you've got other ideas, or know where you'd think to look for these, please share!

**Comment by helderco** [comment]
In Europa we used release notes in GH:

- https://github.com/dagger/dagger/pull/2019

It's something anyone that sends a PR can control, but it's tied to a goreleaser release.

In Python, I'm used to look first at a project's doc site for release notes (e.g., [Django](https://docs.djangoproject.com/en/4.1/releases/), [Wagtail](https://docs.wagtail.org/en/stable/releases/index.html)). If it doesn't exist, then I look for a repo's `CHANGELOG.md`. So this makes sense for versioned releases of the Python library to PyPI. It should be the same with JS/TS but not with Bash nor necessarily with CUE.

At least for the core API though (and any other that makes sense), I really like the locked issue that you can subscribe to, and you have a rich editing experience like you said. So 👍 

TLDR; since we'll have multiple releases for different components we can mix and match what's more appropriate for each. But using conventional commits and at least giving a short `BREAKING CHANGE` notice there with `!` tag (as in #2019), could be used across the board as a common denominator.
